Subscriptions

gleono opened the pull request:
It took me a little more than expected, dealing with a repeated property of Keys was a little tricky.
One significant change that I think I saw before submitting this was that the owner property in Stream is the user_id() and not his/her email. (Recommended practice)

There's a bug, where after subscribing/unsubscribing from streams, it takes a while to reflect that change. It might maybe that the browser is caching the page, but after refreshing you can see actual status of user's subscription.

At this point I think we can refactor a bunch of duplicated code and but it in a base class. 
Also is worth considering pushing more querying and updating methods into the models.

Commit:
created stream service (in-work)

Commit:
Subscribed implemented

Commit:
Merge branch 'master' into subscriptions

Commit:
Subscribe button toggles when subscribed.

Commit:
Unsubscribe implemented

Commit:
Unsubscribe implemented

Commit:
Merge branch 'subscriptions' of https://github.com/rayolanderos/UT-APT-MiniProject into subscriptions

Commit:
Fixed accidentally deleted line

Commit:
Streams user is subscribed to shown on manage page

Commit:
Done implementing unsubscribe in manage page.

Commit:
Delete logging

rayolanderos commented:
I believe the "bug" you are describing is a problem of the local datastore. It takes a while to update itself. It happens to me when creating/deleting streams as well. You have to re-refresh to see the resource appear/gone. I've been ignoring it because it also happened in some other test projects that I downloaded to get a better understanding of some features. I suspect this won't be an issue when it's deployed but we should keep an eye out for it.

rayolanderos commented on controllers/create_stream.py:
```diff
@@ -30,13 +30,12 @@ def post(self):
                         'tags': stream_tags,
                         'cover_url': cover_url, 
-                        'owner': owner_email }
+                        'owner': user.user_id() }
```
Ok with this change. I'll need to create a bunch of stuff again so that it's tied to my id and not my email but I forgive you since I was responsible for the use of email in the first place :)

rayolanderos commented on controllers/subscription.py:
```diff
@@ -0,0 +1,55 @@
+        if result.status_code == 200:
+            self.redirect('{}?id={}'.format(self.uri_for('view', _full=True), stream_id))
+        else:
+            logging.error('Impossible to unsubscribe. User: {}, Stream: {}'.format(user, stream_id))
```
Cool.. Use of methods makes everything look cleaner and it's probably better for maintenance.

rayolanderos commented on models/connexus_user.py:
```diff
@@ -1,5 +1,24 @@
+        return cls.query().filter(cls.user_id == str(user_id)).get()
+
+    @classmethod
+    def is_subscribed(cls, user_id, stream_key):
```
This reminds me I need a get method for the report to show properly in the Trending screen.

rayolanderos commented on services/manage.py:
```diff
@@ -5,30 +5,33 @@
+                    'name': stream.name, 
+                    'date': date_string, 
+                    'photo_count': len(stream.photos),
+                    'all_time_views': len(stream.views_list)
```
All time views should not be the views_list since entries in this list are purged every hour with the cron. There is another field in the model called "views" which reflects the all time views. 

Crons don't run locally, you have to manually run it to test it, but if you do, you'll notice most of the items in the view_list will be removed.

gleono replied:
And this is why we have code reviews.

rayolanderos replied:
I LOVE this feature. I am so sad I have no one at work to do this with, haha

rayolanderos reviewed (changes requested):
Looks great! My only request is to review the all time counter for manage.

gleono commented:
Oh yeah, I wasn't sure which to use and it was already late. I'll fix it when I get home.

gwoodard3 commented on controllers/manage.py:
```diff
@@ -16,50 +18,75 @@
-            method=urlfetch.POST,
-            headers= {'Content-Type': 'application/json'}
-        )
+            result = urlfetch.fetch(
```
This messed me up a few times too, where if you put a blank line in a loop/if statement, it wouldn't indent it.

gleono replied:
Yeah indentation in Python could be a little too special and some editors are not really good. Which one are you using?

gwoodard3 replied:
I'm using PyCharm right now

gwoodard3 reviewed (approved):
I'm not seeing anything wrong myself

Commit:
Fixing a bug where the user_id and the email were registered as new Connexus users

gleono commented:
Done, I changed to use `stream.views`. Also noticed that I introduced a bug by not putting the `try_create_connexus_user()` in the right place. Is fixed now too.

rayolanderos reviewed (approved):
Looks great :)
